globus_cli/commands/ls.py

```diff
@@ -101,19 +38,53 @@ def ls_command(endpoint_plus_path, recursive_depth_limit,
     client = get_client()
     autoactivate(client, endpoint_id, if_expires_in=60)
 
-    # get the `ls` result
-    # note that `path` can be None
-    res = _get_ls_res(client, path, endpoint_id, recursive,
-                      recursive_depth_limit, all)
+    # create the query paramaters to send to operation_ls
+    ls_params = {"show_hidden": int(all)}
+    if path:
+        ls_params["path"] = path
 
-    def cleaned_item_name(item):
-        return item['name'] + ('/' if item['type'] == 'dir' else '')
+    # get the `ls` result
+    if recursive:
+        res = client.recursive_operation_ls(
+            endpoint_id, depth=recursive_depth_limit, **ls_params)
+    else:
+        res = client.operation_ls(endpoint_id, **ls_params)
+
+    def formated_item_name(item):
+        # determine if the formated item name should end in / or not
+        final_slash = ("/" if item["type"] == "dir" else "")
+
+        # ignore any trailing slashes for easier logic
+        if path and path[-1] == "/":
+            formatted_path = path[:-1]
+        else:
+            formatted_path = path
+
+        # if we are doing a recursive ls we want to display the relative path
+        # from the start of the ls.
+        if recursive:
+
+            # if the ls command has a starting path,
+            # give everything relative to that path
+            if formatted_path:
+                relative_path = re.match(u".*{}/(.*{})".format(
+                    re.escape(formatted_path), re.escape(item["name"])),
+                    item["path"]).group(1)
+            # otherwise give everything after the first /~/ or /
+            else:
+                relative_path = re.match(u"(/~)*/(.*{})".format(
+                    re.escape(item["name"])), item["path"]).group(2)
```

Comment:
I'm looking at these regexes again, and I'm not yet convinced that they're right/safe.

Questions:
- Why start the recursive case with `.*`? What could appear before `formatted_path`, and why can it not, itself, contain multiple matches for `formatted_path`?
- What is the `.*` before `item["name"]` in the recursive case? Isn't the formal case for this that there's nothing? When does that match anything?
- In the non recursive case, why do any regex alteration at all?
- How will this handle S3 (today) and Google Drive (future) paths which contain literal `~` chars?